Add workshop on explaining code (#318)
CodeYourFuture/CYF-Workshops: explaining-code/README.md

> # Explaining code

Reviewer: some learning objectives here please

We often need to explain our code to others. Some examples of when we need to do this:
1. When someone else is going to start working on it with us.
2. When trying to get help with our code from someone else.
3. In a job interview.

Pick a project you've been working on recently to explain some of. The project doesn't need to be complete, but it should be at least 100 lines of code and should take more than a couple of sentences to explain.

## The big picture (15 minutes)

> Most of the time it's useful to start by introducing the big picture of the code (what is the project, what problem does it solve, how does the program generally work) before digging into specifics (e.g. what a function does, what some lines do).

Reviewer: ```suggestion
Start by introducing the big picture of the code:
- What is the project? 
- What problem does it solve? 
- How does the program generally work? 

Do this before digging into specifics (e.g. what a function does, what some lines do).
```

> 5 minutes: On your own, think about what order you would want to explain the big picture ideas of a project. What is it useful for someone to know? What order is it useful to introduce those ideas? Once you have an idea, try thinking through how you would explain the project you've chosen.

Reviewer: ```suggestion
5 minutes: On your own, think about how you will explain the big picture of a project. What is useful for someone to know? In what _order_ is it useful to introduce those ideas? Once you have an idea, try thinking through how you would explain the project you've chosen.
```

Volunteers: Facilitate this discussion:

10 minutes: Discuss as a whole class your ideas for how to introduce the big picture of a project. At the end of this discussion, you should have a template of a script with the different topics you would explain.

## Explaining specific code (15 minutes)

> 5 minutes: On your own, think about how you would like a new codebase to be explained to you? Some things to consider:

Reviewer: ```suggestion
5 minutes: On your own, think: how would you like a new codebase to be explained to you? Some things to consider:
```

* Where would you want the explainer to start?
* Would you want to go depth first (keep following all of the function calls before you move into the next line) or breadth first (read each line of a function, before looking at the functions it calls)?
* What kind of example data may be useful to consider?

> * Is it more interesting to be focusing on the normal working cases or unusual cases (e.g. errors)?

Reviewer: ```suggestion
* Is it more interesting to focus on normal cases or unusual cases (e.g. errors)?
```

# Explaining code

## Learning objectives

```objectives
- [ ] Make a strategy for how to explain code.
- [ ] Explain the big picture (what not how) of a software system.
- [ ] Explain how functions and lines of code in a software system work.
```

We often need to explain our code to others. Some examples of when we need to do this:
1. When someone else is going to start working on it with us.
2. When trying to get help with our code from someone else.
3. In a job interview.

> Pick a project you've been working on recently to explain some of. The project doesn't need to be complete, but it should be at least 100 lines of code and should take more than a couple of sentences to explain.

Reviewer: ```suggestion
Pick a project you've been working on recently to explain. The project doesn't need to be complete, but it should be at least 100 lines of code and should take more than a couple of sentences to explain.
```
